### PR TITLE
jsonrpc: PVR.ChannelSwitch optimization.

### PR DESCRIPTION
**Pull request by FSSDawid**
Don't open a new stream if the pvr is in playback just switch to a channel instead.

### CODE REVIEW

**Comment by nemphys** [comment]
Looks good, I just think Lars will moan about the indentation, please fix

**Comment by FSSDawid** [comment]
@nemphys grrrr thanks :)

**Comment by opdenkamp** [comment]
@FSSDawid more moaning from me: please squash these into one commit and do a forced push to this branch. and it's optimisation/optimization :)

**Comment by FSSDawid** [comment]
Well a response to your moaning :) forced pushed ;) 
Regards

**Comment by opdenkamp** [comment]
this needs another change too:
you can only perform a channel switch when you're switching to something of the same type.
so radio -> radio and tv -> tv. you can't do that when a recording is playing.

**Comment by FSSDawid** [comment]
rebased squashed fixed waiting for mo moaning <lool> ;-)

**Comment by FSSDawid** [comment]
Can I switch to a recording while a recording is playing  ? I have assumed we can do that if not just moan :D

**Comment by opdenkamp** [comment]
no you can't. you can only switch channels when live tv or radio is playing.
also, instead of the bunch of if-then-else's, just use CPVRManager::GetCurrentChannel().
if it returns true, a channel is playing. then you can compare the IsRadio() value from the current channel with the value of the new channel. if it's the same, you can switch channels. if not, you have to open it in the standard way.

**Comment by FSSDawid** [comment]
Should be ok now.

File: xbmc/interfaces/json-rpc/PVROperations.cpp [hunk 1]
```diff
@@ -49,7 +49,25 @@ JSON_STATUS CPVROperations::ChannelSwitch(const CStdString &method, ITransportLa
     CLog::Log(LOGDEBUG, "JSONRPC: switch to channel '%d'", iChannelId);
 
     const CPVRChannel *channel = g_PVRChannelGroups->GetByChannelIDFromAll(iChannelId);
-    return g_PVRManager.StartPlayback(channel) ? ACK : InternalError;
+
+    CPVRChannel currentChannel;
+    if (g_PVRManager.GetCurrentChannel(currentChannel))
+    {
+      if (currentChannel.IsRadio() && channel->IsRadio() || !currentChannel.IsRadio() && !channel->IsRadio())
```
Comment (opdenkamp) [inline xbmc/interfaces/json-rpc/PVROperations.cpp]: if (currentChannel.IsRadio() == channel->IsRadio())

Comment (FSSDawid) [inline xbmc/interfaces/json-rpc/PVROperations.cpp]: done

**Comment by opdenkamp** [comment]
you also have to check whether g_PVRChannelGroups->GetByChannelIDFromAll() doesn't return NULL, or you'll get a crash when a channel id is provided that does not or no longer exist.

**Comment by FSSDawid** [comment]
added:
-    if (channel == NULL)
-      return InternalError;

**Comment by FSSDawid** [comment]
Done in 3 lines :)  you are a monster :D lool ;)

**Comment by opdenkamp** [comment]
not really, but I just rather not merge things in that isn't clean or can be done better. if I don't ask you to do it now, I'll have to do it myself later when pvr will be merged with mainline xbmc.

anyway, merged. thanks :)

**Comment by FSSDawid** [comment]
> anyway, merged. thanks :)
> No problem ;) I will get used to it in a while :)  

Regards

**Pull request merged**